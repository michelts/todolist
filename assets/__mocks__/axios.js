export default {
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
};
