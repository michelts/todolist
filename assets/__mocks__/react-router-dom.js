export const useHistory = jest.fn();

export const Link = () => null;
