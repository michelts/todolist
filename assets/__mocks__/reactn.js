export const useGlobal = jest.fn();
