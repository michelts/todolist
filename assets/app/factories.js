import Factory from 'rosie';

const UserFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('username', (index) => `username_${index}`)
  .sequence('name', (index) => `Name-${index}`);

export {
  UserFactory,
};
