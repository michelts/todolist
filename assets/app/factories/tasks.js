import { Factory } from 'rosie';

const TaskFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('description', (index) => `Task ${index}`)
  .attr('priority', 1)
  .attr('due_date', '2020-05-22')
  .attr('is_completed', false);

export {
  TaskFactory,
}
