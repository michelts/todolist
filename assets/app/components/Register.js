import React from 'react';
import { useGlobal } from 'reactn';
import axios from 'axios';
import { useHistory } from 'react-router-dom';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';
import Form from 'react-bootstrap/Form';
import Alert from 'common/components/Alert';
import Input from 'common/components/Input';
import PasswordInput from 'common/components/PasswordInput';

export const RegisterSchema = Yup.object().shape({
  username: Yup.string().required(),
  username: Yup.string().required(),
  password: Yup.string().required(),
});

const Register = () => {
  const history = useHistory();
  const [saveFailed, setSaveFailed] = React.useState(false);
  const [, setUser] = useGlobal('user');

  const initialValues = {
    username: '',
    name: '',
    password: '',
  };

  const handleSubmit = (values) => {
    axios.post('/api/v1/users/', values)
      .then(({ data }) => {
        setUser(data);
        history.push('/tasks');
      })
      .catch(() => setSaveFailed(true));
  };

  return (
    <>
      <h1>Create an account</h1>

      {saveFailed && (
        <Alert>Registration failed! Check the filled data.</Alert>
      )}

      <Formik
        initialValues={initialValues}
        onSubmit={handleSubmit}
      >
        {({ handleSubmit: formikSubmit }) => (
          <Form onSubmit={formikSubmit}>
            <Field component={Input} name="username" label="Username" />

            <Field component={Input} name="name" label="Username" />

            <Field component={PasswordInput} name="password" label="Password" />

            <button type="submit">Submit</button>
          </Form>
        )}
      </Formik>
    </>
  );
};

export default Register;
