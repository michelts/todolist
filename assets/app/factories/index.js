import { UserFactory } from './user';
import { TaskFactory, BlankTaskFactory } from './tasks';

export {
  UserFactory,
  TaskFactory,
  BlankTaskFactory,
};
