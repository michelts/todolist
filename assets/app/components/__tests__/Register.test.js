import React from 'react';
import { useGlobal } from 'reactn';
import { shallow } from 'enzyme';
import axios from 'axios';
import { useHistory } from 'react-router-dom';
import { Formik } from 'formik';
import { UserFactory } from '../../factories';
import Register from '../Register';

const getWrapper = () => {
  const history = {
    push: jest.fn(),
  };
  useHistory.mockReturnValue(history);
  const wrapper = shallow(<Register />);
  return { wrapper, history };
};

const formikChild = (wrapper) => {
  const handleSubmit = jest.fn().mockName('formikSubmit');
  return wrapper.find(Formik).renderProp('children')({ handleSubmit });
};

describe('Register component', () => {
  let setUser;

  beforeEach(() => {
    jest.spyOn(React, 'useEffect').mockImplementation((effect) => effect());
    setUser = jest.fn();
    useGlobal.mockReturnValue([undefined, setUser]);
  });

  it('should render a registration form', async () => {
    const { wrapper } = getWrapper();
    expect(wrapper).toMatchSnapshot('Formik');
    expect(formikChild(wrapper)).toMatchSnapshot('Formik children');
  });

  it('should create the user and redirect him to the task list when form is submitted successfully', async () => {
    const loggedInUser = UserFactory.build();
    axios.post.mockResolvedValue({ data: loggedInUser });

    const { wrapper, history } = getWrapper();
    const payload = { username: 'foo', password: 'bar', name: 'Foo Bar' };
    wrapper.find(Formik).simulate('submit', payload);
    expect(axios.post).toHaveBeenCalledWith('/api/v1/users/', payload);

    await axios.post;
    expect(history.push).toHaveBeenCalledWith('/tasks');
    expect(setUser).toHaveBeenCalledWith(loggedInUser);
  });

  it('should indicate creation failed if form submission is not successful', async (done) => {
    axios.post.mockRejectedValue({});

    const { wrapper, history } = getWrapper();
    const payload = { username: 'foo', password: 'bar', name: 'Foo Bar' };
    wrapper.find(Formik).simulate('submit', payload);
    expect(axios.post).toHaveBeenCalledWith('/api/v1/users/', payload);

    await axios.post;
    expect(history.push).not.toHaveBeenCalled();

    process.nextTick(() => {
      expect(wrapper).toMatchSnapshot('Error msg and Formik');
      expect(formikChild(wrapper)).toMatchSnapshot('Formik children');
      done();
    });
  });
});
