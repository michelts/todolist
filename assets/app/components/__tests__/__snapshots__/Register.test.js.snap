// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component should indicate creation failed if form submission is not successful: Error msg and Formik 1`] = `
<Fragment>
  <h1>
    Create an account
  </h1>
  <Alert>
    Registration failed! Check the filled data.
  </Alert>
  <Formik
    initialValues={
      Object {
        "name": "",
        "password": "",
        "username": "",
      }
    }
    onSubmit={[Function]}
  >
    <Component />
  </Formik>
</Fragment>
`;

exports[`Register component should indicate creation failed if form submission is not successful: Formik children 1`] = `
<Form
  inline={false}
  onSubmit={[MockFunction formikSubmit]}
>
  <Field
    component={[Function]}
    label="Username"
    name="username"
  />
  <Field
    component={[Function]}
    label="Username"
    name="name"
  />
  <Field
    component={[Function]}
    label="Password"
    name="password"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</Form>
`;

exports[`Register component should render a registration form: Formik 1`] = `
<Fragment>
  <h1>
    Create an account
  </h1>
  <Formik
    initialValues={
      Object {
        "name": "",
        "password": "",
        "username": "",
      }
    }
    onSubmit={[Function]}
  >
    <Component />
  </Formik>
</Fragment>
`;

exports[`Register component should render a registration form: Formik children 1`] = `
<Form
  inline={false}
  onSubmit={[MockFunction formikSubmit]}
>
  <Field
    component={[Function]}
    label="Username"
    name="username"
  />
  <Field
    component={[Function]}
    label="Username"
    name="name"
  />
  <Field
    component={[Function]}
    label="Password"
    name="password"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</Form>
`;
